lib/ImplicitDiscreteSolve/src/cache.jl

```diff
@@ -1,36 +1,66 @@
-mutable struct ImplicitDiscreteState{uType, pType, tType}
+struct ImplicitDiscreteState{uType, pType, tType}
     u::uType
     p::pType
-    t_next::tType
+    t::tType
 end
 
-mutable struct IDSolveCache{uType} <: OrdinaryDiffEqMutableCache
+mutable struct IDSolveCache{uType, cType, thetaType} <: OrdinaryDiffEqMutableCache
     u::uType
     uprev::uType
-    state::ImplicitDiscreteState
-    prob::Union{Nothing, SciMLBase.AbstractNonlinearProblem}
+    z::uType
+    nlcache::cType
+    Θks::thetaType
 end
 
 function alg_cache(alg::IDSolve, u, rate_prototype, ::Type{uEltypeNoUnits},
         ::Type{uBottomEltypeNoUnits}, ::Type{tTypeNoUnits}, uprev, uprev2, f, t,
         dt, reltol, p, calck,
         ::Val{true}) where {uEltypeNoUnits, uBottomEltypeNoUnits, tTypeNoUnits}
-    state = ImplicitDiscreteState(isnothing(u) ? nothing : zero(u), p, t)
-    IDSolveCache(u, uprev, state, nothing)
+    state = ImplicitDiscreteState(zero(u), p, t)
+    f_nl = (resid, u_next, p) -> f(resid, u_next, p.u, p.p, p.t)
+
+    u_len = length(u)
+    nlls = !isnothing(f.resid_prototype) && (length(f.resid_prototype) != u_len)
+    prob = if nlls
+        NonlinearLeastSquaresProblem{isinplace(f)}(
+            NonlinearFunction(f_nl; resid_prototype = f.resid_prototype),
+            u, state)
+    else
+        NonlinearProblem{isinplace(f)}(f_nl, u, state)
+    end
+
+    nlcache = init(prob, alg.nlsolve)
+
+    IDSolveCache(u, uprev, state.u, nlcache, uBottomEltypeNoUnits[])
 end
 
 isdiscretecache(cache::IDSolveCache) = true
 
-struct IDSolveConstantCache <: OrdinaryDiffEqConstantCache
-    prob::Union{Nothing, SciMLBase.AbstractNonlinearProblem}
-end
+# struct IDSolveConstantCache <: OrdinaryDiffEqConstantCache
+#     prob::Union{Nothing, SciMLBase.AbstractNonlinearProblem}
+# end
```

Comment:
This is dead code on master. Is there any benefit in adding this variant in first place?

Reply:
imo dead code should be deleted. If we want it again, that's what git history is for.